### ISSUE TITLE
Bumpline - bumping into zeros -- add checks for uniqueness of combinations.

### ISSUE THREAD

**Issue by ericmelse** [asjadnaqvi/stata-bumpline#6]
Dear Asjad,

I continued with some of my other (survey) data using `bumpline` as it is supposed to work.
But, I run into a problem which, I assume, is caused because of the presence of zeros in the (summed) values to be ranked relatively.
Before I show you my results, let me introduce you first to the data set.
A survey was administered to students from different colleges who volunteer to participate in a program that will prepare them for the next level of schooling (i.e. now they are at a college and next fall they might enroll in university).
The data set as such is rather small: N=275.
The bumpline plot that I have in mind draws their choice of topic that they will follow during the program:
```
-------------------------------------------------------------------
                      |      Freq.    Percent      Valid       Cum.
----------------------+--------------------------------------------
Valid   1 Orienteren  |        144      52.36      52.36      52.36
        2 Analyseren  |         43      15.64      15.64      68.00
        3 Samenwerken |         64      23.27      23.27      91.27
        4 Reflecteren |         24       8.73       8.73     100.00
        Total         |        275     100.00     100.00           
-------------------------------------------------------------------
```
So, these should become the 'bumping lines' and is set by using `by(varname)`
Next, is the categorical variable to use for the x-axis, which are age groups:
```
--------------------------------------------------------------------
                       |      Freq.    Percent      Valid       Cum.
-----------------------+--------------------------------------------
Valid   1 17 of jonger |         51      18.55      18.55      18.55
        2 18 jaar      |        103      37.45      37.45      56.00
        3 19 jaar      |         57      20.73      20.73      76.73
        4 20 jaar      |         25       9.09       9.09      85.82
        5 21 jaar      |         28      10.18      10.18      96.00
        6 Onbekend     |         11       4.00       4.00     100.00
        Total          |        275     100.00     100.00           
--------------------------------------------------------------------
```
And the (categorical) variable to use for the y-axis, are three participating colleges where the program is given:
```
-------------------------------------------------------------------------
                            |      Freq.    Percent      Valid       Cum.
----------------------------+--------------------------------------------
Valid   1 Nova College      |         15       5.45       5.45       5.45
        2 Horizon College   |         99      36.00      36.00      41.45
        3 ROC van Amsterdam |        161      58.55      58.55     100.00
        Total               |        275     100.00     100.00           
-------------------------------------------------------------------------
```
So, my expectation is a rather straightforward bumpline plot that draws four lines (topics), each of which is ranked relatively by college (y) over age groups (x). Or stated by my research question: what is the ranked difference between colleges and between age groups for the expressed interest in topics? Should there be small differences in the ranking order between age groups then we can expect a rather horizontal pattern. So, my command is:
`bumpline c_roc c_leeftijd , by(KEUZEDEELKEUZE) xsize(2) ysize(1) lw(*1) msym(S) ms(2.3) mc(white) name(roc, replace)`
which produces this result:
![Bumpline_test_ROC_c_roc_c_leeftijd_20230414](https://user-images.githubusercontent.com/3686136/231996219-e3dd70e8-115e-4afa-b9b2-bfff1b846f17.png)
What seems to be calculated **correctly** is:
- The four bump lines (topics)
- The use of the age groups for the sequential order along the x-axis
- The use of the colleges for the relative rank order along the y-axis

What seems to be calculated **not correctly** is:
- The drawing of the bump lines in the sequential order along the x-axis: a whole bunch or (apparently) erratic positions are created (?) in between x-axis order positions
- The use and drawing of the relative rank along the y-axis: your 'minimum' of 10 (the default) is used although there is nothing else present in the data than the 4 colleges.

Next, I tried using the `top(num)` option to settle to 4, using:
`bumpline c_roc c_leeftijd , by(KEUZEDEELKEUZE) xsize(2) ysize(1) lw(*1) msym(S) ms(2.3) mc(white) top(4) name(rocT, replace)`
which produces this result:
![Bumpline_test_ROC_c_roc_c_leeftijd_Top4_20230414](https://user-images.githubusercontent.com/3686136/231999246-90e9930d-b980-4f4f-a79d-db4a36bcffdf.png)
It seems that the relative order is calculated such that their position is 'fixed' to the default 10 instead the expected 4 and using `top(4)` 'chops' the y-axis as such but also has a bearing on what is happening in the calculation of the relative order itself. Do compare this visualization with the top 4 positions of the bump lines in the previous visualization.

So, why all these problems?

My gist is that `bumpline` 'bumps into zeros' and is not able to 'handle zeros' as a relative ordered result. The reason for this is that your code appears to assume that the 'numeric' y-variable should be data scaled and not be categorical.  Therefore,  `bumpline` seems to expect that always there is (summed) data to be used to rank order relatively without any zeros.

In my example, the summed data to be used for calculating the relative ranking of the 4 topics does contain zeros (obviously).
Do inspect their tabulation:
`bysort KEUZEDEELKEUZE: tab c_roc c_leeftijd`

for 1 Orienteren
```
      Partner MBO |                            c_leeftijd
       Keuzedelen | 17 of jon    18 jaar    19 jaar    20 jaar    21 jaar   Onbekend |     Total
------------------+------------------------------------------------------------------+----------
     Nova College |         1          3          2          1          1          0 |         8 
  Horizon College |        11         25         17          5          2          0 |        60 
ROC van Amsterdam |        16         26         18          5          8          3 |        76 
------------------+------------------------------------------------------------------+----------
            Total |        28         54         37         11         11          3 |       144 
```
for 2 Analyseren
```
      Partner MBO |                            c_leeftijd
       Keuzedelen | 17 of jon    18 jaar    19 jaar    20 jaar    21 jaar   Onbekend |     Total
------------------+------------------------------------------------------------------+----------
     Nova College |         0          2          1          0          0          0 |         3 
  Horizon College |         3          5          1          3          2          0 |        14 
ROC van Amsterdam |         5          8          1          4          5          3 |        26 
------------------+------------------------------------------------------------------+----------
            Total |         8         15          3          7          7          3 |        43 
```
for 3 Samenwerken
```
      Partner MBO |                            c_leeftijd
       Keuzedelen | 17 of jon    18 jaar    19 jaar    20 jaar    21 jaar   Onbekend |     Total
------------------+------------------------------------------------------------------+----------
     Nova College |         0          2          1          0          0          0 |         3 
  Horizon College |         2         11          4          0          2          1 |        20 
ROC van Amsterdam |        11         13          9          4          4          0 |        41 
------------------+------------------------------------------------------------------+----------
            Total |        13         26         14          4          6          1 |        64 
```
for 4 Reflecteren
```
      Partner MBO |                            c_leeftijd
       Keuzedelen | 17 of jon    18 jaar    19 jaar    20 jaar    21 jaar   Onbekend |     Total
------------------+------------------------------------------------------------------+----------
     Nova College |         0          1          0          0          0          0 |         1 
  Horizon College |         0          2          0          0          1          2 |         5 
ROC van Amsterdam |         2          5          3          3          3          2 |        18 
------------------+------------------------------------------------------------------+----------
            Total |         2          8          3          3          4          4 |        24 
```

So, in short,  `bumpline` should be enabled to rank sort relatively also when the y variable is a categorical variable.
Should you want to use my data and code to replicate and use it for development, I certainly can share that with you.

Best,
Eric

**Comment by asjadnaqvi** [comment]
Dear Eric,

thank you finding this! If you can share the data, it will be very helpful to fix this bug.

best,
Asjad

**Comment by ericmelse** [comment]
Dear Asjad,

I attach the file used for the above examples:

`use "Case_categorical_bumpline_with_zeros.dta" , clear`

Later today I can respond, should you have follow up questions.

[Case_categorical_bumpline_with_zeros.zip](https://github.com/asjadnaqvi/stata-bumpline/files/11238444/Case_categorical_bumpline_with_zeros.zip)

**Comment by asjadnaqvi** [comment]
Dear Eric,

This is not a valid `bumpline` dataset but more of an [alluvial](https://github.com/asjadnaqvi/stata-alluvial) plot. `bumpline` needs each `y` `x` variable combination to be unique for each `by()` group and here we have a lot of cross tabs. I will add a check for this.

Either one prepares it such that it fits the `bumpline` criteria, or just use `alluvial`:

```
alluvial c_leeftijd c_roc KEUZEDEELKEUZE
```

![alluvial](https://user-images.githubusercontent.com/38498046/232619840-6d9ea3dc-d502-4cbf-88e1-3cd6e410945c.png)

**Comment by ericmelse** [comment]
Dear Asjad,
Yes, that is very good suggestion and I started immediately to explore `alluvial` to this purpose. I have already posted some of my suggestions after my first experiences with it on your GitHub page of `alluvial`.
I do disagree a very little bit with you that my example data as such is better not used to create a bumpline plot. But, I do agree that I am asking, possibly, too much from `bumpline` to do this work for me. The issue really is to be able to 'count' the zeros and treat them as instances of a 'true score' and thus be able to sum these and use that to sort them along `y` with the other response categories over the `x-axis ordered categories` (time, state or group). Basically, it is about a different (simplified) view on how (in this example) each groups differ in their preferences). That can be visualized with other plots (e.g. stacked bars) but I prefer the bumped line plot for that purpose.

**Comment by ericmelse** [comment]
My first result - working with what `alluvial` now has to offer - now is:
![Alluvial_test_ROC_20230418](https://user-images.githubusercontent.com/3686136/232854690-aa90c7d1-ceb1-46fd-8f4d-5081931aa29e.png)
The one thing that I cannot set are the level titels, so I did that using Stata's editor of the graphic window to tweak their size, like:
![Alluvial_test_ROC_edit_20230418](https://user-images.githubusercontent.com/3686136/232855000-3f2f437d-9091-48d9-a8d5-69ecd2de0d03.png)
but other (regular) label options are beyond my control.
Note that had to 'hack' the category labels rather extensively to bypass the issues that I have already discussed on the alluvial github page.
Most exotic is the 'trick' to get such labels (as well as their value) positioned over multiple 'lines'. I learned about that on the [Statalist here](https://www.statalist.org/forums/forum/general-stata-discussion/general/1417586-including-line-break-in-variable-labels).
So, my brutal coding now is:
```
use "$pathfile/MBO22Keuzedelen_Start_20230413.dta" , clear
gen c_keuzedl=1
replace c_keuzedl=2 if KEUZEDEELKEUZE==3
replace c_keuzedl=3 if KEUZEDEELKEUZE==2
replace c_keuzedl=4 if KEUZEDEELKEUZE==4
label define lbl_c_keuzedl 1 "Oriënteren          `=char(13)'`=char(10)'   " 2 "Samenwerken          `=char(13)'`=char(10)'     " 3 "Analyseren           `=char(13)'`=char(10)'    " 4 "Reflecteren          `=char(13)'`=char(10)'    " , replace
label values c_keuzedl lbl_c_keuzedl

label define lbl_vrouw 0 "Mannen `=char(13)'`=char(10)'" 1 "Vrouwen `=char(13)'`=char(10)' " 2 "Neutraal" , replace
label values vrouw lbl_vrouw

label define lbl_c_roc 1 "        Nova" 2 "    Horizon`=char(13)'`=char(10)'    " 3 "          ROC van Am-`=char(13)'`=char(10)'         sterdam" , replace
label values c_roc lbl_c_roc
```
Because I now cannot control the color coding of the levels individually, I have tried a good deal of Ben Jann's palettes.
One of them - palette(tab20c) - gives me an acceptable result (see above). The command used is:
`
alluvial c_roc vrouw c_keuzedl , laba(0) labs(*1.4) showtot boxw(*80) shares vals(*1.2) valf(%12.2f) valgap(*18) palette(tab20c) name(roc2, replace)
`
